Parse excel file

pva701 opened the pull request:
## Description

What has been done:
- [x] old format parsing implemented
- [x] servant machinery defined (including endpoint to upload a file)
- [x] initial config defined
- [x] set up swagger schema 
- [x] simple frontend to show upload results

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-12

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

pva701 committed:
Add parser and dummy API

Problem: we need to upload an excel file with experiment.

Solution: parse excel file on backend and add endpoint
to return parsed data.

pva701 committed:
Add config and config parser

Problem: we need to specify some server parameters,
like port and whether we need to generate swagger docs.

Solution: add a YAML config and parse it with Data.Yaml

pva701 committed:
Add servant server runner

Problem: we need to run servant server.

Solution: add machinery to run it.

gromakovsky commented on backend/src/Edna/Web/API.hs:
```diff
@@ -0,0 +1,34 @@
+     ( EdnaAPI
+     , EdnaEndpoints (..)
+     , ednaAPI
+     ) where
```
Formatting of exports is inconsistent and doesn't match style guide in most modules :cry:

pva701 replied:
It was a test to check your attention (to check that you are alive at all, I think ability to being able to notice such things determine your state being alive).

I will reformat to 2 spaces everywhere in the code.

gromakovsky commented on backend/edna.cabal:
```diff
@@ -30,11 +30,29 @@ common common-options
       src
+  build-depends:
+    aeson,
+    aeson-options,
```
`aeson-options` is [deprecated](https://github.com/serokell/aeson-options), please let's not use it, let's use [aeson-casing](https://hackage.haskell.org/package/aeson-casing) instead.

pva701 replied:
I just wanna notice how Jonn and Arseniy are just having fun together in this PR :putya:

gromakovsky commented on backend/app/Main.hs:
```diff
@@ -4,7 +4,28 @@ module Main
+    fullDesc <> progDesc "Edna API server."
+
+  Y.decodeFileEither @EdnaConfig configPath >>= \case
+    Left e -> putText (show e)
```
In this case we will exit with 0 code which is not what we want. Also the error message should probably go to stderr. Why not use `decodeFileThrow` which does the right thing we want here?

gromakovsky commented on backend/src/Edna/Config.hs:
```diff
@@ -0,0 +1,32 @@
+data ApiConfig = ApiConfig
+  { acListenAddr :: NetworkAddress
+  , acServeDocs :: Bool
+  } deriving stock (Generic, Read, Show)
```
What do we need `Read` for?

gromakovsky commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,90 @@
+      ( \case
+          CellDouble x -> Right x
+          _ -> Left UnexpectedCellType
+      )
```
Apparently you always pass the same lambda to `readSheet'`, can you extract it into a named function?

gromakovsky commented on backend/stack.yaml:
```diff
@@ -2,3 +2,8 @@ resolver: lts-17.2
+extra-deps:
+  - aeson-options-0.1.0
+
+allow-newer: true
```
Is it needed for `aeson-options` (that we shouldn't use because it's deprecated) or something else?

pva701 replied:
yes, I will remove if it's needed for `aeson-casing`

gromakovsky commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,90 @@
+
+uploadExperiment :: MultipartData Mem -> Handler [ExperimentalMeasurement]
+uploadExperiment multipart = do
+  let file = U.head (files multipart)
```
Is it guaranteed to be non-empty?

gromakovsky commented:
> tests for parsing

Note that the format will change, so don't put too much effort into it. I mean, I think it still makes sense to add tests now, so that we have necessary code in place and can simply update test files later. But let's have just one sample file for now.

pva701 committed:
Add swagger schema

Problem: we would like to generate client types
and API requests automatically by swagger schema.

Solution: serve swagger docs from servant.

gromakovsky commented:
> simple frontend to show upload results

Note that it's a separate issue, [EDNA-13](https://issues.serokell.io/issue/EDNA-13).

gromakovsky commented on backend/src/Edna/Web/Error.hs:
```diff
@@ -0,0 +1,45 @@
+instance Buildable ExperimentParsingError where
+  build UnexpectedCellType = "Expected one type of cell but got another one"
+  build InvalidCell = "Undefined content of the cell"
+  build (WorksheetNotFound w) = "" +| w |+ " worksheet not found"
```
Is `"" +| ` really needed here?

gromakovsky commented on backend/config.yaml:
```diff
@@ -0,0 +1,3 @@
+api:
```
What do you think about also having default config so that one can run `edna-server` without thinking about config and still get something meaningful?

pva701 replied:
well, I think we can use default config if a config file is not found or something.

sashasashasasha151 commented on backend/src/Edna/Web/API.hs:
```diff
@@ -0,0 +1,34 @@
+
+-- | API type specification.
+type EdnaAPI =
+  "api" :> "v1" :> ToServant EdnaEndpoints AsApi
```
What for we need `"api" :> "v1"`?

pva701 replied:
perhaps, it's redundant indeed

sashasashasasha151 commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,99 @@
+module Edna.Web.Handlers
```
I propose to think about structure of the app from the very beginning\
I know that it is MVP and we do not need to think about unrelated things but I think that even in MVP it is important\
We had this problem in blend app and I don't want to meet it here when you can't understand what is going on and your handler file has 1000+ lines and you want to rewrite all backend\
So, lets split all features to different parts. For example here we have part that works with files and it looks like a separate thing\
Also, I propose to use MVC model from start\
First layer will be DB, it will just make queries and transform DB types to domain types.\
Second layer is for business logic, it only works with domain types\
Third layer is for API which just works with business logic and construct types for response\
@gromakovsky what do you think about it?\
@pva701 I do not suggest you to think about it in this PR as at least we do not have DB now (but I already see that it is possible to split handler on two modules), but I also want to here your opinion about it

gromakovsky replied:
I think this module should have almost no logic and just define `ednaHandlers` with simple definitions. `eeUploadExperiment = bla-bla` where `bla-bla` is something small and uses logic from another module. So in this case I think `parseExperimentXls` should be in a different module, somewhere outside of `Edna.Web`. I think @pva701 just didn't care much about the structure of modules here, which is ok for this PR (as you said).

Overall I agree with this 3-layer approach, but currently there is just no code to apply it :)

pva701 replied:
good point, though, I can't separate this uploading logic from the rest of (future) logic (perhaps because there is no logic yet and I can't come up with it). so I more or less agree but don't want to do it in this PR :putya:

gromakovsky replied:
What about just creating a new module and putting `parseExperimentXls` there?

gromakovsky replied:
Whatever, we'll do it along with implementing proper parser (now that we have example of inputs).

sashasashasasha151 commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,90 @@
+
+-- | Server handler implementation for Edna API.
+ednaHandlers :: EdnaHandlers Handler
+ednaHandlers =
+  genericServerT
+    EdnaEndpoints
```
```suggestion
ednaHandlers = genericServerT EdnaEndpoints
```

pva701 replied:
btw, it's how Ormolu in idea formatted it :putya: I will fix

sashasashasasha151 commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,90 @@
+uploadExperiment :: MultipartData Mem -> Handler [ExperimentalMeasurement]
+uploadExperiment multipart = do
+  let file = U.head (files multipart)
+  putStrLn $ "Excel file name " ++ show (fdFileName file)
```
What about adding logs (and also log api calls)?

pva701 replied:
what about not to do ALL the initialisation work in completely unrelated task? we can have this un-merged for a while until all this stuff will be initialised.

feel free to create a task to add it.

sashasashasasha151 commented on backend/src/Edna/Web/Handlers.hs:
```diff
@@ -0,0 +1,90 @@
+      -> [ExperimentalMeasurement]
+    mergeCells compounds concentrations signals =
+      mapMaybe
+        (\(rc, cmp) -> ExperimentalMeasurement cmp <$> M.lookup rc concentrations <*> M.lookup rc signals)
```
Too long line

pva701 replied:
like @jagajaga's nick ? :putya:

sashasashasasha151 commented on backend/app/Main.hs:
```diff
@@ -4,7 +4,28 @@ module Main
 import Universum
 
-import Stub (stub)
+import qualified Data.Yaml as Y
```
Why not using `Loot`?

pva701 replied:
because it's piece of overcomplicated shit. Do you even know why loot config was created and why do we need to use it at all? (I really doubts)

also I tried to use it with current resolver and there are a lot of compile time errors. feel free to add it yourself :putya:

sashasashasasha151 commented:
Also @pva701, why you don't add task name to commits and branch name?

gromakovsky commented:
https://github.com/serokell/edna/pull/1#issuecomment-776822123

pva701 committed:
Add dummy fronted to upload an excel table with measurements

Problem: we need to check what has been uploaded to a server.

Solution: add dummy table to show content of excel file parsed by server.

sashasashasasha151 commented on frontend/.eslintrc.js:
```diff
@@ -20,6 +20,9 @@ module.exports = {
     "@typescript-eslint/no-explicit-any": 0,
 
+    "react/jsx-closing-bracket-location": 0,
+    "react/no-array-index-key": 0,
```
https://stackoverflow.com/questions/46735483/error-do-not-use-array-index-in-keys \
Please do not remove it

pva701 replied:
what you suggest using instead? Keys are a pain in the ass.

sashasashasasha151 replied:
If it is just a dummy component you can use `index.toString()`\
What index we will use later? Maybe generate indexes on backend or you can also generate it on frontend. The problem is that if we will do some changes in this table or somehow filter/sort rows we will see how it doesn't work)

sashasashasasha151 commented on frontend/.eslintrc.js:
```diff
@@ -20,6 +20,9 @@ module.exports = {
     // Sometime we need to specify any type, so it's rather useful
     "@typescript-eslint/no-explicit-any": 0,
 
+    "react/jsx-closing-bracket-location": 0,
```
Why?

pva701 replied:
because construction
```
<Comp
    attr1=..
    attr2=..
>
  .. code here
</Comp>
```

looks ugly, what is the point of moving > at the separate line? it's not improve readability to me. does it do for you? it just looks like nonsense to me. why should we keep this rule?

sashasashasasha151 replied:
I don't think that it is ugly as it helps you to quickly separate `attrs` from `code here`, but if it really important for you I am ok with it

sashasashasasha151 commented on frontend/src/utils/utils.ts:
```diff
@@ -0,0 +1,4 @@
+// eslint-disable-next-line import/prefer-default-export
+export function isDefined<T>(anyVal: T | undefined | null): anyVal is T {
```
Where are you using it?

pva701 replied:
to avoid writing `x != null && x != undefined` whenever I want to check it. 
(don't suggest if (x) because empty strings, lists, zero values will be implicitly converted to false in this case)

what you suggest using instead?

pva701 replied:
oh, yes, nowhere. perhaps worth removing

sashasashasasha151 commented on frontend/src/pages/upload/UploadPage.tsx:
```diff
@@ -0,0 +1,80 @@
+          e.preventDefault();
+          try {
+            if (uploadInputRef.current
+          && uploadInputRef.current.files
+          && uploadInputRef.current.files.length > 0) {
```
Our config is bad) \
Will take a look later

sashasashasasha151 commented on frontend/src/components/Spinner.scss:
```diff
@@ -0,0 +1,61 @@
+// Taken from here https://projects.lukehaas.me/css-loaders/
```
It is just for example, right?

pva701 replied:
yes, just random filler to show it's uploading

pva701 committed:
Address review #6 issues

gromakovsky reviewed (approved):
Don't want to delay merging further.

sashasashasasha151 commented on README.md:
```diff
@@ -23,6 +23,10 @@ There is nothing to build yet.
 then you can connect to it via `psql` or using any convenient database viewer.
 
+To run a server run `stack exec -- edna-server -c config.yaml`:
+* The server endpoints available at `localhost:9000/api/v1/*`.
```
Remove v1
